### DISCUSSION TITLE
Exchange order

### DISCUSSION THREAD

**Discussion by lsarni** [urql-graphql/urql#3647] [Q&A] [answered]
We are having some glitches on how the data is loaded, and I wonder if we might not have our exchanges in the correct order. I think it would be really useful to document the correct order in a single place. I see on the [authentication docs](https://commerce.nearform.com/open-source/urql/docs/advanced/authentication/) _The authExchange is an asynchronous exchange, so it must be placed in front of all fetchExchanges but after all other synchronous exchanges, like the cacheExchange._ but then it's not always stated if an exchange is async or sync, or where should it go on the list. 

In particular we have
- devtoolsExchange
- authExchange
- requestPolicyExchange
- offlineExchange
- fetchExchange

Is this order correct?

**Comment by lsarni** [comment]
I just found this explaining the order, but I'm not sure I know which ones are sync/asycn https://commerce.nearform.com/open-source/urql/docs/advanced/authoring-exchanges/#synchronous-first-asynchronous-last

**Comment by JoviDeCroock** [comment, marked as answer]
In your case you want

```
devtoolsExchange <-- debugging
requestPolicyExchange <-- synchronous changes to the context
offlineExchange <-- synchronous access to the cache
authExchange <-- asynchronous, can do requests to renewAuth/...
fetchExchange <-- asynchronous, reaches out to your back-end
```

**Reply by lsarni** [reply]
Thanks for the quick reply!

Our `offlineExchange` uses `AsyncStorage` https://commerce.nearform.com/open-source/urql/docs/graphcache/offline/#react-native does that make it async?

**Reply by JoviDeCroock** [reply]
No rehydration happens in the background. A cache has a synchronous need as you don't want to delay load-unload by a micro-tick due to implicit Promises.

**Reply by lsarni** [reply]
Thank you again :)
